server/game/cards/22-BtB/TheMadKingsCommand.js

```diff
@@ -0,0 +1,142 @@
+                },
+                attachment: {
+                    choosingPlayer: 'each',
+                    optional: true,
+                    ifAble: true,
+                    messages: Messages.eachPlayerTargetingForCardType('attachments'),
+                    numCards: 3,
+                    activePromptTitle: 'Select up to 3 attachments',
+                    cardCondition: card => card.location === 'play area' && card.getType() === 'attachment'
+                },
+                location: {
+                    choosingPlayer: 'each',
+                    optional: true,
+                    ifAble: true,
+                    messages: Messages.eachPlayerTargetingForCardType('locations'),
+                    numCards: 3,
+                    activePromptTitle: 'Select up to 3 locations',
+                    cardCondition: card => card.location === 'play area' && card.getType() === 'location'
+                },
+                hand_shadows: {
+                    choosingPlayer: 'each',
+                    optional: true,
+                    ifAble: true,
+                    messages: Messages.eachPlayerSecretTargetingForCardType('cards in hand and/or shadows'),
+                    numCards: 3,
+                    activePromptTitle: 'Select up to 3 cards',
+                    cardCondition: (card, context) => (card.location === 'hand' || card.location === 'shadows') && card.controller === context.choosingPlayer
+                }
+            },
+            handler: context => {
+                this.promptPlayersForOrder(context.targets.selections);
+            }
+        });
+    }
+
+    promptPlayersForOrder(selections) {
+        let toMove = [];
+        let validInPlayTypes = ['character', 'attachment', 'location'];
+
+        for(let selection of selections) {
+            let player = selection.choosingPlayer;
+            let selectedCards = selection.value || [];
+            let selectionTargets = selection.name.split('_');
+            // Selections (above) can be grouped by underscores. This splits those groups into their own comparisons
+            for(let selectionTarget of selectionTargets) {
+                let targetLocation = validInPlayTypes.includes(selectionTarget) ? 'play area' : selectionTarget;
+                let cardsInLocation = this.game.allCards.filter(card => card.controller === player 
+                    && card.location === targetLocation // checking location
+                    && (validInPlayTypes.includes(selectionTarget) ? card.getType() === selectionTarget : true) // checking type, if applicable
+                    && card.allowGameAction('returnCardToDeck')); // checking return to deck
+                let playerSpecificToMove = cardsInLocation.filter(card => !selectedCards.includes(card));
+                toMove = toMove.concat(playerSpecificToMove);
+            }
+        }
```

Comment:
The location and type logic, while likely correct, is hard to follow. 

I'm wondering if this would be more straightforward if you gathered all of the cards of certain types / locations, then removed the chosen ones from that list. E.g. something like (but maybe not exactly)
```javascript
const potentialCards = this.game.allCards.filter(card => (
    card.allowGameAction('returnCardToDeck') && (
        ['hand', 'shadows'].includes(card.location) || 
        card.location === 'play area' && ['character', 'location', 'attachment'].includes(card.getType())
    )
));
const selectedCards = selections.reduce((selectedCards, selection) => {
    return selectedCards.concat(selection.value || []);
}, []);
const toMove = potentialCards.filter(card => !selectedCards.includes(card));
```

Reply:
Solid suggestion! The above code makes more sense to me; I've implemented & tested it, and it works as intended. Code has been pushed in e0a7c45.

It was so complicated as the selections kept changing throughout development, and I was getting tired of adjusting it. 😄